l1-contracts/contracts/upgrades/L1GenesisUpgrade.sol

```diff
@@ -0,0 +1,106 @@
+                    (_chainId, _forceDeploymentsData)
+                ); //todo
+                complexUpgraderCalldata = abi.encodeCall(
+                    IComplexUpgrader.upgrade,
+                    (L2_GENESIS_UPGRADE_ADDR, l2GenesisUpgradeCalldata)
+                );
+            }
+            // complexUpgraderCalldata = abi.encodeCall(ISystemContext.setChainId, (_chainId));
+
+            // slither-disable-next-line unused-return
+            (, uint32 minorVersion, ) = SemVer.unpackSemVer(SafeCast.toUint96(_protocolVersion));
+            l2ProtocolUpgradeTx = L2CanonicalTransaction({
+                txType: SYSTEM_UPGRADE_L2_TX_TYPE,
+                from: uint256(uint160(L2_FORCE_DEPLOYER_ADDR)),
+                // to: uint256(uint160(L2_SYSTEM_CONTEXT_SYSTEM_CONTRACT_ADDR)),
+                to: uint256(uint160(L2_COMPLEX_UPGRADER_ADDR)),
+                gasLimit: PRIORITY_TX_MAX_GAS_LIMIT,
+                gasPerPubdataByteLimit: REQUIRED_L2_GAS_PRICE_PER_PUBDATA,
+                maxFeePerGas: uint256(0),
+                maxPriorityFeePerGas: uint256(0),
+                paymaster: uint256(0),
+                // Note, that the protocol version is used as "nonce" for system upgrade transactions
+                nonce: minorVersion,
+                value: 0,
+                reserved: [uint256(0), 0, 0, 0],
+                data: complexUpgraderCalldata,
+                signature: new bytes(0),
+                factoryDeps: L2ContractHelper.hashFactoryDeps(_factoryDeps),
+                paymasterInput: new bytes(0),
+                reservedDynamic: new bytes(0)
+            });
+        }
+        ProposedUpgrade memory proposedUpgrade = ProposedUpgrade({
+            l2ProtocolUpgradeTx: l2ProtocolUpgradeTx,
+            factoryDeps: _factoryDeps,
+            bootloaderHash: bytes32(0),
+            defaultAccountHash: bytes32(0),
+            verifier: address(0),
+            verifierParams: VerifierParams({
+                recursionNodeLevelVkHash: bytes32(0),
+                recursionLeafLevelVkHash: bytes32(0),
+                recursionCircuitsSetVksHash: bytes32(0)
+            }),
+            l1ContractsUpgradeCalldata: new bytes(0),
+            postUpgradeCalldata: new bytes(0),
+            upgradeTimestamp: 0,
+            newProtocolVersion: _protocolVersion
+        });
+
+        Diamond.FacetCut[] memory emptyArray;
+        Diamond.DiamondCutData memory cutData = Diamond.DiamondCutData({
+            facetCuts: emptyArray,
+            initAddress: _l1GenesisUpgrade,
+            initCalldata: abi.encodeCall(this.upgradeInner, (proposedUpgrade))
+        });
+        Diamond.diamondCut(cutData);
+
+        emit GenesisUpgrade(address(this), l2ProtocolUpgradeTx, _protocolVersion, _factoryDeps);
+        return Diamond.DIAMOND_INIT_SUCCESS_RETURN_VALUE;
+    }
+
+    function upgradeInner(ProposedUpgrade calldata _proposedUpgrade) external override returns (bytes32) {
```

Comment:
why do we need upgradeInner instead of just upgrading?

Reply:
because of Calldata vs memory